### ISSUE TITLE
Exception ordering with explicit NULLCHKs 

### ISSUE THREAD

**Issue by 0xdaryl** [eclipse/openj9#6347]
Many (many, many) years ago, the IA32 JIT fixed a problem where a `TR::ResolveAndNULLCHK` of a call/load/store implemented with an **explicit NULLCHK** was not operating correctly because the resolution and NPE exceptions would be generated in the wrong order.  This was because the explicit NULLness checks would wrap the call/load/store and any NPE would occur before any resolution exception which violates the VM spec.  I don't recall what the test case was, but I believe it was a JCK test.

If the NULLness check is done implicitly this won't be a problem--I think--because the resolution will happen first before the indirect address child is dereferenced (i.e., the implicit NULLCHK).

The original author of the NULLCHK evaluator code knew of this problem which is why he left comments like this in the code, which still exist in Power and ARM:
```
   // TODO - If a resolve check is needed as well, the resolve must be done
   // before the null check, so that exceptions are handled in the correct
   // order.
   //
   ///// if (needsResolve)
   /////    {
   /////    ...
   /////    }
```
As far as I can tell, this problem was not solved on Power, Z, ARM, or AMD64 (IA32 is OK).  Can @fjeremic @gita-omr please confirm that this is still a problem, or describe why it is not for the respective platform?  Thanks.

FYI:  @knn-k @aarongraham9 @andrewcraik

**Comment by fjeremic** [comment]
The `resolveCHKEvaluator` on Z doesn't actually do anything [1]. It's effectively a pass-through evaluator because it's the `isUnresolved` flag on the child that determines whether resolution is needed and the resolution logic is generated when the child is evaluated.

Looking at the `evaluateNULLCHKWithPossibleResolve` in OMR though [2] (this should really migrate to OpenJ9) I can see paths on which we may be evaluating the NULL check reference node and carrying out the NULL check before evaluating the first child of the `TR::ResolveAndNULLCHK` which the resolve is acting on.

This evaluator is _very_ confusing though and poorly written so it's hard to trace all the paths by hand. I can't say for certain whether there is a problem without a set of unit tests.

[1] https://github.com/eclipse/openj9/blob/fdf59d0b5f27a303f8e1fba2b57d03b433b2895f/runtime/compiler/z/codegen/J9TreeEvaluator.cpp#L3014-L3042
[2] https://github.com/eclipse/omr/blob/8a77e0c833cbfb7409d0ee2f5780107d8b3f2d18/compiler/z/codegen/ControlFlowEvaluator.cpp#L1992-L2428

**Comment by 0xdaryl** [comment]
Thanks @fjeremic.  @gita-omr, any thoughts from the Power perspective?

**Comment by gita-omr** [comment]
Similar to Z, on P, resolve check is implicit and is done by evaluating the child. So I think this comment can be removed on Power.

**Comment by fjeremic** [comment]
> Similar to Z, on P, resolve check is implicit and is done by evaluating the child. So I think this comment can be removed on Power.

I don't think this is the case. As far as I can see on Power in `evaluateNULLCHKWithPossibleResolve`  we always generate the NULLCHK before the resolve [1], so almost certainly the bug described by @0xdaryl exists on Power today.

The NULLCHK is generated at line in [2] and is unguarded. There is a call to `cg->evaluate` a few lines before which evaluates the reference which the NULLCHK is acting on. Note that this reference may not be the same as the node the resolveCHK is acting on. We only evaluate the first child, i.e. the one we need to resolve _after_ the NULLCHK at the bottom of the evaluator in [3].

I think the bug described in the OP exists on Power.

[1] https://github.com/eclipse/omr/blob/d51a93b2b5c9256b0417092173d271e67b0ce6c9/compiler/p/codegen/ControlFlowEvaluator.cpp#L3875-L3980
[2] https://github.com/eclipse/omr/blob/d51a93b2b5c9256b0417092173d271e67b0ce6c9/compiler/p/codegen/ControlFlowEvaluator.cpp#L3912-L3915
[3] https://github.com/eclipse/omr/blob/d51a93b2b5c9256b0417092173d271e67b0ce6c9/compiler/p/codegen/ControlFlowEvaluator.cpp#L3970-L3975

**Comment by fjeremic** [comment]
@0xdaryl if I understand correctly for the following tree:

```
n1n  ResolveAndNULLCHK on n3n
n2n    call <someMethod>
n3n      aload <bla>
```

The resolve is acting on the child and the NULLCHK on the grandchild. In this case the bug that was fixed was that the resolve had to happen first and then the NULLCHK? Am I understanding this correctly? I think a tree example may help.
